[8.0][l10n_es_aeat_sii][IMP] Mejoras en la usabilidad

pedrorgil opened the pull request:
Añadidas varias mejoras de cara a usabilidad:
- Se cambia el permiso para ver la información SII del formulario de factura del grupo. Cambiada de 'Gestor Financiero' al grupo 'Cobros y pagos' de contabilidad. Creo que es coherente que quienes pueden crear y modificar facturas puedan ver y rellenar esta información, si no habrá casos en los que se les requiera un campo (por ejemplo el tipo de rectificativa) que no puedan ni ver.
- Se asigna permiso sólo al grupo 'Responsable AEAT' para ver el resultado del envío al SII y los trabajos del conector en el formulario de factura.
- Se corrige un error detectado en el formulario de facturas de proveedor, no se establecía readonly el campo 'Descripción SII' para el método de relleno 'auto'. Por error está sólo contemplado en el formulario de factura de cliente, no en las de proveedor.
- Correcciones en la programación de hora de envío para el conector por el método 'Hora fija'. Se evitan errores por no estar establecido 'tz' en el contexto (ahora se coge directamente del timezone del usuario) y se corrige la obtención de los minutos (que no estaba funcionando y obtenía la hora programada pero los minutos siempre eran 0).

pedrorgil committed:
[8.0][l10n_es_aeat_sii][IMP] Mejoras en la usabilidad y corrección de errores

pedrorgil commented:
He detectado también algunas traducciones que faltan de los últimos commit. @pedrobaeza las meto aquí o se hacen por otro lado?

pedrorgil commented:
Vale me autocontesto, ya lo he hecho por transifex

pedrobaeza commented on l10n_es_aeat_sii/models/res_company.py:
```diff
@@ -65,10 +65,10 @@ class ResCompany(models.Model):
         if self.send_mode == 'fixed':
-            offset = datetime.now(pytz.timezone(
-                self._context.get('tz'))).strftime('%z')
+            tz = pytz.timezone(self.env.user.partner_id.tz)
```
El usuario puede tener la tz también vacía. Lo primero sería coger correctamente la tz de esta forma:
```
tz = self.env.context.get('tz', self.env.user.partner_id.tz)
```
Después, simplemente hacer el offset si hay tz.

pedrobaeza commented on l10n_es_aeat_sii/wizards/account_invoice_refund.py:
```diff
@@ -41,6 +41,14 @@ def _selection_sii_refund_type(self):
 
+    @api.onchange('filter_refund')
+    def onchange_filter_refund(self):
+        if self.sii_refund_type_required:
```
Puedes reducir la complejidad ciclomática así:
```
if not self.sii_refund_type_required:
    return
self.sii_refund_type = 'I' if self.filter_refund == 'refund' else 'S'
```

pedrobaeza reviewed (changes requested):
Cambia la versión

pedrorgil commented:
@pedrobaeza perfecto, cambios hechos.

omar7r commented:
LGTM

pedrobaeza commented:
Esto habría que complementarlo con permisos extendidos para el conector:

- Permiso de lectura a queue.job para todos los que pueden visualizar la factura y ven el campo "Trabajos del conector". Se puede añadir regla de seguridad para limitarlo a los trabajos del canal correspondiente.
- Crear el trabajo con sudo, pero pasando el usuario que lo lanza como propietario para evitar errores de permisos al validar la factura.

Si no puedes hacerlo tú ahora, lo prepararé yo luego en otro PR juntando todo luego esta noche.

pedrorgil commented:
Hasta mañana no podré hacerlo, como veas, si quieres/puedes hazlo tú y si no lo hago yo mañana

pedrorgil commented:
Creo que también hay que dar permisos de lectura sobre account.tax.template al grupo account.group_account_invoice ya que es necesario para poder acceder al mapeo de impuestos para generar el fichero. Ahora mismo no lo puedo comprobar, lo digo para que lo tengas en cuenta si lo haces tú

pedrobaeza committed:
[IMP] l10n_es_aeat_sii: Seguridad + usabilidad

* Filtro de no enviados SII que sólo incluya 2017
* Los trabajos no realizados se eliminan en lugar de pasarlos a estado "Hecho"
  al reenviar, cancelar, etc.
* Permiso de lectura a queue.job para los gestores de la AEAT.
* Regla de seguridad para que los gestores de la AEAT sólo vean los trabajos
  del SII.
* Los trabajos se encolan como admin para evitar errores al validar por gente
  sin permisos del conector. Desafortunadamente, no se puede mantener el
  usuario original en el trabajo.

pedrobaeza commented:
He incluido en este PR los cambios mencionados:

* Filtro de no enviados SII que sólo incluya 2017
* Los trabajos no realizados se eliminan en lugar de pasarlos a estado "Hecho" al reenviar, cancelar, etc.
* Permiso de lectura a queue.job para los gestores de la AEAT.
* Regla de seguridad para que los gestores de la AEAT sólo vean los trabajos del SII.
* Los trabajos se encolan como admin para evitar errores al validar por gente sin permisos del conector. Desafortunadamente, no se puede mantener el usuario original en el trabajo.

Me faltaría, que lo hago mañana:

* Poner una vista de los trabajos del conector más adecuada, con los campos relevantes, como fecha de ejecución).
* Añadir un botón en esa vista para poder empezar el trabajo inmediatamente, para cubrir esta casuística:
  * Se tiene un retardo de envío grande configurado.
  * Cuando llega el envío, tiene un error.
  * Se corrige y se vuelve a reenviar, ¡pero ese trabajo se programa con ese mismo retardo configurado y ya no va a entrar en tiempo!

  Aquí barajaba la posibilidad de hacer ese reenvío inmediato, pero eso supone cambiar interfaz para mostrar el botón adecuado cuando corresponde y también parte de la lógica, así que creo que ésta es la mejor solución. Opinión?

pedrobaeza commented:
Se me olvidaba comentar que no hace falta ese permiso de account.tax.template porque el trabajo ya se ejecuta como admin y no lo va a necesitar.

pedrorgil commented:
El botón de envío inmediato lo veo bien. Lo que podríamos hacer es enviar inmediatamente si ha fallado el envío, se supone que si ya se ha intentado enviar no lo vas a querer encolar otra vez, para esto no hace falta otro botón solo controlarlo en el envio.

pedrorgil commented:
El filtro para las pendientes de envio SII no puede hacerse con date_invoice, debe hacerse por periodo. Puedes tener facturas de proveedor de hace meses imputadas a partir de Julio. Estoy haciendo algunas mejoras, en un rato hago commit

pedrobaeza commented:
@pedrorgil llevas razón que puedo ver si la factura tiene sii_send_failed y cambiar el ETA en torno a eso. Ya lo he implementado. He añadido también el botón para envío immediato y el campo en la vista de los trabajos.

Sobre lo del periodo, llevas razón, pero entonces no vamos a poder ponerlo, porque la nomenclatura de periodos y las fechas de inicio/fin pueden variar por implantación y entonces que no sea efectivo.

Cuidado con sobreescribir mi commit. Haz git pull. Qué tienes planeado? Tengo que llevar esto a v9 ya mismo, así que cuánto tiempo tardarías en meterlo?

pedrorgil commented:
SI ya he hecho pull, si vas a meter más cosas me espero y lo hago de nuevo antes del commit.
Estoy metiendo un campo compute con store a False y un search que nos permita ver cuándo debe mostrarse el botón: el estado de la factura, el date_start del periodo es a partir de 01-07-2017, no hay trabajos pendientes, la posición fiscal está activada para el SII y el estado de envío al SII. Con todas esas variables se calcula un Booleano que meto en la vista y en función de el se muestra o no el botón. El search lo estoy implementando para habilitar las búsquedas, lo hago por SQL que es más rápido y así no afecta casi al rendimiento.
¿Como lo ves?

pedrobaeza commented:
No, en principio ya no metería nada más, pero cuidado que he hecho `git push -f` en este último para juntar mi commit. No entiendo lo de mostrarse el botón. Te refieres al filtro?

pedrorgil commented:
No, me refiero a no mostrar el botón 'Enviar SII' si ya está enviada o hay programado un envío, el usuario se puede quedar 'Pillado' si le da a 'Enviar SII', le pega el pantallazo y sigue estando el botón. Es probable que le de un par de veces por si acaso :)
Ademas voy a meter la opción 'Cancelar envío' en la lista de trabajos, por si se quiere cancelar un envío momentaneamente sin tener que cancelar la factura. Por ejemplo si ya la has validado y no tienes claro que esté bien y quieres parar el envío de momento.

pedrobaeza commented:
Bueno, ahora lo que ocurre al darle de nuevo a "Enviar a SII" es que se borra el anterior trabajo y se encola uno nuevo. No lo veo del todo mal, pero si complica mucho la lógica, pues lo dejamos, que estamos añadiendo mucha complicación. 

Lo del botón de cancelar no está mal, pero 2 cosas:

* Que el botón sea un icono para no ocupar tanto.
* No sé si te dará un error. Otras veces que he puesto un botón que ejecuta el unlink tengo un error de respuesta.

pedrobaeza commented:
Otra cosa: ¿lo ponemos a partir del 1 de enero de 2017 o del 1 de julio de 2017? Técnicamente las del primer semestre se deben enviar también (aunque sea de otra manera). Yo puse del 01/01/2017, pero no sé qué será mejor.

pedrorgil commented:
Yo lo pondría a partir de 01-07-2017, al menos de momento hasta que tengamos implementado los envíos de primer semestre (como le den y se envíe se van a enviar mal). Luego podemos hacer algo a parte para esto, por no complicar más la lógica de los envíos normales para un caso que una vez hecho ya no se va a hacer más, no se algo tipo wizard que procese todo y un control de las fallidas con reintento, porque nadie va a entrar a enviar una a una las facturas del primer semestre no?

pedrobaeza commented:
Uhm, pero pueden dar error y tener que reenviar algunas... Vale, de momento dejamos desde Julio y luego vemos lo otro.

pedrorgil commented:
He incluido la lógica de comprobación de envíos pendientes. No es sencilla, si ves que complica mucho la lógica podemos quitarlo, aunque a mi me gusta tener esto controlado que luego los usuarios la lían sin darse cuenta...

pedrobaeza commented:
@pedrorgil la lógica para mostrar el botón se hace demasiado complicada para lo poco que aporta a nivel de usabilidad, ya que si se pulsa el botón de nuevo cuando está visible (que aún no se ha comunicado al SII la factura o la modificación), lo único que se hace es reencolar el trabajo, y puesto que ahora tenemos botones para hacer que se ejecute inmediatamente o de cancelar el envío (que lo he rescatado de tu commit), entonces no merece la pena.

He dejado el PR con los cambios a excepción de esto último y paso a integrarlo también en la rama principal.

pedrorgil commented:
Ok a mí tampoco me acababa de convencer la complicación que mete para lo que se consigue, sobre todo de cara a mantenimiento  en futuros cambios, pero me lié a hacerlo y ya que estaba lo metí para que lo valoráramos más de uno. Me parece bien no incluirlo

pedrorgil commented:
Revisando tu commit se te ha pasado meter el botón de cancelar envio en las facturas de proveedor. Lo metes tu o hago PR mañana?

pedrobaeza commented:
Es cierto. Esto de llevar dos vistas separadas es proclive a error. Se tendría que crear una vista única para queue.job con prioridad baja y hacer que se utilice esa vista ahí mediante el context. Ahora preparo el cambio y lo pongo por commit directo para dejarlo bien.

pedrobaeza commented:
Ya he puesto que compartan vista en https://github.com/OCA/l10n-spain/commit/bbbc5a633ed25bef9794b2ea1011001786e33c68 para que así no se vuelva a olvidar más cualquier cambio que se haga en un sitio y no en el otro. No he cambiado nº de versión como esta tarde ya lo he cambiado.

pedrorgil commented:
👍
